### PR TITLE
Added json ouput of snapshots

### PR DESCRIPTION
**Pull request by ibib**
This PR adds json support as requested in [#726](https://github.com/restic/restic/issues/726)

### CODE REVIEW

File: src/cmds/restic/cmd_snapshots.go [hunk 1]
```diff
@@ -26,6 +27,7 @@ The "snapshots" command lists all snapshots stored in a repository.
 type SnapshotOptions struct {
 	Host  string
 	Paths []string
+	Json  bool
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Make this a global option, so we can add it to other commands as well.

File: src/cmds/restic/cmd_snapshots.go [hunk 2]
```diff
@@ -87,6 +86,22 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
 
 	}
 
+	if opts.Json == true {
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: The `==` is unneeded, please remove it.

File: src/cmds/restic/cmd_snapshots.go [hunk 3]
```diff
@@ -87,6 +86,22 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+		jsonoutput(list)
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Add a `return` after this and remove the `else` branch.

File: src/cmds/restic/cmd_snapshots.go [hunk 4]
```diff
@@ -87,6 +86,22 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+	} else {
+		naturaloutput(list)
+	}
+
+	return nil
+}
+
+//naturaloutput provides human redability
+func naturaloutput(list []*restic.Snapshot) {
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: This function needs a new name, all functions in the restic binary share the same namespace. What about `printSnapshots()`?

File: src/cmds/restic/cmd_snapshots.go [hunk 5]
```diff
@@ -121,5 +136,40 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+	return
+}
+
+//Snapshot provides struct to store snapshots
+type Snapshot struct {
+	ID        string   `json:"id"`
+	Date      string   `json:"date"`
+	Host      string   `json:"host"`
+	Tags      []string `json:"tags"`
+	Directory []string `json:"directory"`
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Please also use the plural `Directories` here.

File: src/cmds/restic/cmd_snapshots.go [hunk 6]
```diff
@@ -121,5 +136,40 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+}
+
+//jsonoutput provides machine redability
+func jsonoutput(list []*restic.Snapshot) error {
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: This function also needs a different name. What about `printSnapshotsJSON()`?

File: src/cmds/restic/cmd_snapshots.go [hunk 7]
```diff
@@ -121,5 +136,40 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+			ID:        sn.ID().Str(),
+			Date:      sn.Time.Format(TimeFormat),
+			Host:      sn.Hostname,
+			Tags:      sn.Tags,
+			Directory: sn.Paths}
+
+		response = append(response, k)
+	}
+
+	output, _ := json.Marshal(response)
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Use a `json.Decoder` here, like `dec := json.NewDecoder(os.Stdout)` and then use its `Decode()` method.

**Review by fd0 — CHANGES_REQUESTED** [review]
Thanks for your work, please see my review comments.

**Comment by ibib** [comment]
Thank you for your annotations.
Sorry that this PR is a bit ugly, this is my first opensource contribution.

I am working now on the JSON encode thingy and will then look in how to resolve conflicts.

The describing text at the global JSON option was choosen on purpose as it only works with the snapshot option at the moment. `print snapshots in json format`

I cant use `printSnapshots()` as function name as it is already in use [here](https://github.com/restic/restic/blob/master/src/cmds/restic/cmd_forget.go#L63).
Would this be a candidate for refactoration?

**Comment by fd0** [comment]
> Sorry that this PR is a bit ugly, this is my first opensource contribution.

Oh, that's great, I'm honored!

> The describing text at the global JSON option was chosen on purpose as it only works with the snapshot option at the moment. print snapshots in json format

Okay, we'll leave it for now.

> I cant use printSnapshots() as function name as it is already in use here.
> Would this be a candidate for refactoration?

Yes, but not in this PR.

I'll add further comments now.

Are you aware that you accidentally merged the master branch in again? I know that git and GitHub are a hard to understand for beginners.

File: src/cmds/restic/cmd_snapshots.go [hunk 8]
```diff
@@ -87,6 +84,25 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+			fmt.Fprintf(os.Stderr, "error printing snapshot: %v\n", err)
+		}
+		return nil
+	}
+	printSnapshotsReadable(list)
+
+	return nil
+}
+
+//printSnapshotsReadable provides human redability
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Please describe what the function does, e.g.:

```
// printSnapshotsReadable prints a text table of the snapshots in list to stdout.
```

Also add a space character after the comment (`//`), and make sure that the description is a complete sentence, as described in https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.

File: src/cmds/restic/cmd_snapshots.go [hunk 9]
```diff
@@ -132,5 +148,49 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
 
 	tab.Write(os.Stdout)
 
+	return
+}
+
+//Snapshot provides struct to store snapshots
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Actually I'd prefer to use the JSON representation we're storing in the repository, it is defined on the `Snapshot` struct here (so you do not need a special struct at all): https://github.com/restic/restic/blob/master/src/restic/snapshot.go#L13-L27

File: src/cmds/restic/cmd_snapshots.go [hunk 10]
```diff
@@ -132,5 +148,49 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+type Snapshot struct {
+	ID          string   `json:"id"`
+	Date        string   `json:"date"`
+	Host        string   `json:"host"`
+	Tags        []string `json:"tags"`
+	Directories []string `json:"directories"`
+}
+
+//printSnapshotsJSON provides machine redability
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Add a space after `//`, then describe what the function does, e.g.:
```
// printSnapshotsJSON writes the JSON representation of list to stdout.
```

File: src/cmds/restic/cmd_snapshots.go [hunk 11]
```diff
@@ -132,5 +148,49 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
...
+func printSnapshotsJSON(list []*restic.Snapshot) error {
+
+	enc := json.NewEncoder(os.Stdout)
+	//set JSON prefix
+	fmt.Fprint(os.Stdout, "[")
```
Comment (fd0) [inline src/cmds/restic/cmd_snapshots.go]: Are you aware that you can just pass a `[]Snapshot` to the encoder? Like this:
```go
err := json.NewEncoder(os.Stdout).Encode(list)
```

The encoder prints the list delimiters (`[` and `]`) automatically.

File: src/cmds/restic/global.go [hunk 12]
```diff
@@ -53,6 +54,7 @@ func init() {
 	f.StringVarP(&globalOptions.PasswordFile, "password-file", "p", "", "read the repository password from a file")
 	f.BoolVarP(&globalOptions.Quiet, "quiet", "q", false, "do not output comprehensive progress report")
 	f.BoolVar(&globalOptions.NoLock, "no-lock", false, "do not lock the repo, this allows some operations on read-only repos")
+	f.BoolVarP(&globalOptions.JSON, "json", "", false, `print snapshots in json format`)
```
Comment (fd0) [inline src/cmds/restic/global.go]: Use double quotes (`"`) for the help text. What do you think about the following help text: `set output mode to JSON for commands that support it`

**Review by fd0 — CHANGES_REQUESTED** [review]

**Comment by fd0** [comment]
Hey, do you have any questions? Is there anything I can do to help you? Or do you just not have any spare time at the moment? :)

**Comment by ibib** [comment]
I tried to use [`err := json.NewEncoder(os.Stdout).Encode(list)`](https://gist.github.com/ibib/f705fb8982bcf0295f9e75f7c22ce3f1) as suggested but i couldn't work out how to print the id's.
The latest commit should provide the same result.

**Comment by fd0** [comment]
You can create a new struct `Snapshot` that [embeds](https://golang.org/ref/spec#Struct_types)  `*restic.Snapshot`, and add an `ID` field like this:
```go
// Snapshot helps to print Snaphots as JSON
type Snapshot struct {
	*restic.Snapshot

	ID string `json:"id"`
}
```

It has all the fields `restic.Snapshot` has, and an additional `ID` field, and you don't even need to repeat all field names!

This new struct can then be used to construct a list of `Snapshot` like this:
```go
	var snapshots []Snapshot
	for _, sn := range list {
		k := Snapshot{
			Snapshot: sn,
			ID:       sn.ID().String(),
		}

		snapshots = append(snapshots, k)
	}
```
Finally, just print this list:
```go
	return json.NewEncoder(os.Stdout).Encode(snapshots)
```

**Comment by ibib** [comment]
Wow. That is totally more easy. 

Thank you very much for your help.

**Review by fd0 — APPROVED** [review]

**Review by fd0 — CHANGES_REQUESTED** [review]
Very good! Now before this can get merged, there are a couple of smaller things that need to be done:
 * Squash your changes into one commit
 * Write a short paragraph for the manual, perhaps titled "Scripting restic", which mentions `--json` and shows an example usage with `snapshots`.

Thanks for your work!

**Pull request closed without merging**